### DISCUSSION TITLE
Explicit fpm Manifest for schema generation and fpm.toml validation

### DISCUSSION THREAD

**Discussion by gnikit** [fortran-lang/fpm#748] [General]
I am looking for the the fpm, manifest so as to generate a JSON schema for validation of `fpm.toml` files in VS Code. Do we store it somewhere either in TOML, Markdown or JSON form? I don't think the HTML version used in [docs](https://fpm.fortran-lang.org/en/spec/manifest.html) will do. I did a quick search in the repo but I couldn't find it.

**Comment by awvwgk** [comment]
I have a Pydantic model for fpm at https://github.com/awvwgk/pyfpm, it probably needs some updates to account for the latest changes in fpm, but should be able to produce the required JSON schema.

**Reply by gnikit** [reply]
Cheers, I remember we chatted about this a while back. I am going over to have a look now.

**Reply by gnikit** [reply]
Do you think it would be worth/be relatively easy to move the generation of a schema as part of the CI workflow in fpm? How do we currently check if a schema is valid internally in fpm?

**Reply by awvwgk** [reply]
There is a check procedure when reading the configuration data from the TOML data structure, which checks for conflicting or unsupported entries. The type checking is done via the TOML Fortran `get_value` interface.

**Reply by gnikit** [reply]
So I suppose we can go about this in 2 ways:
1. Manually update `pyfpm` and use it to generate a global manifest and  a schema
2. Or find a way to automatically parse the `get_value` queries and assemble them into a global manifest

Do you see an alternative? Option 1. might be okay for now but it is bound to be prone to bugs and hard to maintain in the future.

**Reply by awvwgk** [reply]
I don't think option 2 is viable, you need to parse and understand Fortran code like

https://github.com/fortran-lang/fpm/blob/c123a1118f7b3405fb60910ce1c6c4c3ce2d7e0a/src/fpm/manifest/dependency.f90#L207

Also, a refactoring of the fpm manifest reading would for sure break the parser, unless it works on the ASR generated by LFortran or similar.

---

To have fpm export a schema, it is necessary for fpm to actually use this schema to validate the manifest, otherwise it not much better than maintaining a separate schema. See https://fortran-lang.discourse.group/t/3963

**Comment by awvwgk** [comment]
I uploaded a generated schema at https://github.com/awvwgk/pyfpm/releases/download/v0.1.0/pyfpm-0.1.0.json, let me if this is sufficient (this schema can validate all fpm projects on GitHub around October 2021).

**Reply by gnikit** [reply]
Thanks, this will have to do for now. Tracked in https://github.com/fortran-lang/vscode-fortran-support/issues/654

**Reply by gnikit** [reply]
Do you think we should bring this repo to fortran-lang?

**Reply by awvwgk** [reply]
That's something we can do if needed

**Reply by awvwgk** [reply]
Open for discussion in https://github.com/awvwgk/pyfpm/issues/1
